File: src/MessageFormatter.ts
export class MessageFormatter {
  private static delimiter = "\x00";

  public static join(lines: any[]): string {
    return lines
      .map((line): string => {
        if (typeof line === "string") {
          return line;
        } else if (line === undefined) {
          return "";
        } else {
          return JSON.stringify(line);
        }
      })
      .join(this.delimiter);
  }

  public static split(value: string, limit: number = 0): string[] {
    const parts = value.split(this.delimiter);

    if (limit === 0) {
      return parts;
    }

    if (parts.length <= limit) {
      return parts;
    }

    const head = parts.slice(0, limit - 1);
    const tail = parts.slice(limit - 1).join(this.delimiter);

    return [...head, tail];
  }
}

File: ios/Detonator/Sources/Detonator/MessageFormatter.swift
import Foundation

public class MessageFormatter {
    private let detonator: Detonator
    
    private static let delimiter = "\u{0000}"

    init(_ detonator: Detonator) {
        self.detonator = detonator
    }

    public func join(_ lines: [Encodable?]) -> String {
        var parts: [String] = []

        for line in lines {
            if let str = line as? String {
                parts.append(str)
            } else {
                parts.append(detonator.encode(line)!)
            }
        }

        return parts.joined(separator: Self.delimiter)
    }

    public func split(_ value: String, _ limit: Int = 0) -> [String] {
        let parts = value.components(separatedBy: Self.delimiter)
        
        if limit == 0 {
            return parts
        }
        
        if parts.count <= limit {
            return parts
        }
        
        let head = parts.prefix(limit - 1)
        let tail = parts.suffix(from: limit - 1).joined(separator: Self.delimiter)
        
        return Array(head) + [tail]
    }
}
